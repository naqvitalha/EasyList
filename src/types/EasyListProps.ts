export interface EasyListProps {
    aheadCount: number;
    children: React.ReactNodeArray | undefined;
};