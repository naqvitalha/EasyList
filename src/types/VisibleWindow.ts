export interface VisibleWindow {
    paddingTillMin: number;
    minCaptureOffset: number;
    min: number;
    max: number;
    size: number;
};