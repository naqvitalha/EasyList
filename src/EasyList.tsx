import * as React from "react";
import { useState } from "react";
import { ScrollView, View, LayoutChangeEvent } from "react-native";
import { EasyListProps } from "./types/EasyListProps";
import { VisibleWindow } from "./types/VisibleWindow";
import { Queue } from "./ds/Queue";

export function EasyList(props: EasyListProps) {
  const initialHeightTracker: Record<number, number[]> = {};
  const initialWindow: VisibleWindow = {
    paddingTillMin: 0,
    minCaptureOffset: 0,
    min: 0,
    max: 0,
    size: 0
  };
  const [heightTracker] = useState(initialHeightTracker);
  const [visibleWindow, setVisibleWindow] = useState(initialWindow);
  const callback = (index: number, height: number, width: number): void => {
    heightTracker[index] = [height, width];
  };
  return (
    <ScrollView>
      <View style={{ height: 200 }} />
      {getChildren(props, 0, callback)}
    </ScrollView>
  );
}

function getChildren(
  props: EasyListProps,
  firstIndex: number,
  callback: (index: number, height: number, width: number) => void
): React.ReactNodeArray {
  const children = props.children || [];
  const maxIndex = children.length - 1;
  if (maxIndex < 0) {
    return children;
  }
  firstIndex = Math.max(firstIndex, 0);
  const aheadCount = props.aheadCount;
  const min = Math.max(0, firstIndex - aheadCount);
  const max = Math.min(maxIndex, firstIndex + aheadCount);
  const res = new Array<React.ReactNode>(max - min + 1);
  for (let i = min; i <= max; i++) {
    res[i - min] = (
      <View
        onLayout={(event: LayoutChangeEvent) => {
          const height = event.nativeEvent.layout.height;
          const width = event.nativeEvent.layout.width;
          callback(i, height, width);
        }}
      >
        {children[i]}
      </View>
    );
  }
  return res;
}

function updatePool(prevMin: number, prevMax:number, newMin:number, newMax:number, pool: Queue<number>, indexKeyMap: Record<number, number>): void {
  
}

//#if [TEST]
export { getChildren };
//#endif
