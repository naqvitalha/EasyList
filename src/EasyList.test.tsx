import { getChildren } from "./EasyList";
import { Text } from "react-native";
import React from "react";
import { EasyListProps } from "./types/EasyListProps";

function getChildrenDataGen(
    childCount: number,
    aheadCount: number
): EasyListProps {
    const arr = new Array(childCount);
    return { children: arr, aheadCount: aheadCount };
}

function callback(index: number, height: number, width: number) {
    //no-op
}

test("Case: No Children or Undefined", () => {
    const props = getChildrenDataGen(0, 0);
    expect(getChildren(props, 0, callback).length).toBe(0);

    props.children = undefined;
    expect(getChildren(props, 0, callback).length).toBe(0);
});

test("Case: Length Checks", () => {
    const data1 = getChildrenDataGen(1, 0);
    expect(getChildren(data1, 0, callback).length).toBe(1);

    const data2 = getChildrenDataGen(1, 5);
    expect(getChildren(data2, 0, callback).length).toBe(1);

    const data3 = getChildrenDataGen(10, 5);
    expect(getChildren(data3, 0, callback).length).toBe(6);
    expect(getChildren(data3, 5, callback).length).toBe(10);
    expect(getChildren(data3, -1, callback).length).toBe(6);
    expect(getChildren(data3, 2, callback).length).toBe(8);
    expect(getChildren(data3, 8, callback).length).toBe(7);
});

test("Case: Response Format", () => {
    const data1 = getChildrenDataGen(10, 5);
    expect(getChildren(data1, 0, callback)).toMatchSnapshot();
    expect(getChildren(data1, 3, callback)).toMatchSnapshot();

    data1.children = new Array(5).map((item) => {
        return <Text>Test me</Text>;
    });
    expect(getChildren(data1, 8, callback)).toMatchSnapshot();
});
