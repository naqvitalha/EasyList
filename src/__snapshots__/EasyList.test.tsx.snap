// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Case: Response Format 1`] = `
Array [
  <View
    onLayout={[Function]}
  />,
  <View
    onLayout={[Function]}
  />,
  <View
    onLayout={[Function]}
  />,
  <View
    onLayout={[Function]}
  />,
  <View
    onLayout={[Function]}
  />,
  <View
    onLayout={[Function]}
  />,
]
`;

exports[`Case: Response Format 2`] = `
Array [
  <View
    onLayout={[Function]}
  />,
  <View
    onLayout={[Function]}
  />,
  <View
    onLayout={[Function]}
  />,
  <View
    onLayout={[Function]}
  />,
  <View
    onLayout={[Function]}
  />,
  <View
    onLayout={[Function]}
  />,
  <View
    onLayout={[Function]}
  />,
  <View
    onLayout={[Function]}
  />,
  <View
    onLayout={[Function]}
  />,
]
`;

exports[`Case: Response Format 3`] = `
Array [
  <View
    onLayout={[Function]}
  />,
  <View
    onLayout={[Function]}
  />,
]
`;
